### PR TITLE
create compoundAsk API

### PR DESCRIPTION
**Pull request by PeterTheOne**
Needs cleanup/restructuring because of duplicate code. Feedback very welcome! @yaronkoren @JeroenDeDauw

I need this to make https://github.com/SemanticMediaWiki/SemanticMediaWiki ajax functionality work with SemanticCompoundQueries.

### CODE REVIEW

**Comment by JeroenDeDauw** [comment]
I suspect you need to submit this against Wikimedia's Gerrit.

**Comment by PeterTheOne** [comment]
Ok, I have no experience with Gerrit yet, I will look into it.

**Comment by yaronkoren** [comment]
@PeterTheOne - what format(s) would this be used for? And does it work?

File: SCQ_CompoundAsk.php [hunk 1]
```diff
@@ -0,0 +1,143 @@
...
+use SMW\MediaWiki\Api\ApiRequestParameterFormatter;
+
+/**
+ * API module to query SMW by providing multiple queries in the ask language.
+ *
+ * @ingroup SemanticCompoundQueries
+ *
+ * @author Peter Grassberger < petertheone@gmail.com >
+ */
+class SCQCompoundAsk extends Query {
```
Comment (JeroenDeDauw) [inline SCQ_CompoundAsk.php]: Such inheritance is not nice design wise. Might not be entirely avoidable when using the MediaWiki API. Still, you can just derive from the base API class and break out of that context (by using another class or classes) as soon as you can. (Example: https://github.com/wmde/WikibaseQuery/blob/master/src/Api/EntitiesByPropertyValue.php#L42)

Any comment on this @mwjames?

File: SCQ_CompoundAsk.php [hunk 2]
```diff
@@ -0,0 +1,143 @@
...
+    public function execute() {
+        $parameterFormatter = new ApiRequestParameterFormatter( $this->extractRequestParams() );
+        $outputFormat = 'json';
+
+        $parameters = $parameterFormatter->getAskApiParameters();
+
+        /**
+         * from SCQQueryProcessor::doCompoundQuery
+         */
+        $other_params = array();
```
Comment (JeroenDeDauw) [inline SCQ_CompoundAsk.php]: $otherParameters

File: SCQ_CompoundAsk.php [hunk 3]
```diff
@@ -0,0 +1,143 @@
...
+            if ( strpos( $param, '[' ) !== false ) {
+                $queryParams[] = $param;
+            } else {
+                $parts = explode( '=', $param, 2 );
+
+                if ( count( $parts ) >= 2 ) {
+                    $other_params[strtolower( trim( $parts[0] ) )] = $parts[1]; // don't trim here, some params care for " "
+                }
+            }
+        }
```
Comment (JeroenDeDauw) [inline SCQ_CompoundAsk.php]: Looks like this foreach could be in its own method

File: SCQ_CompoundAsk.php [hunk 4]
```diff
@@ -0,0 +1,143 @@
...
+            );
+
+            $results = SCQQueryProcessor::mergeSMWQueryResults( $results, $next_result->getResults() );
+            $printRequests = SCQQueryProcessor::mergeSMWPrintRequests( $printRequests, $next_result->getPrintRequests() );
+        }
+
+        // Sort results so that they'll show up by page name
+        uasort( $results, array( 'SCQQueryProcessor', 'compareQueryResults' ) );
+
+        $query_result = new SCQQueryResult( $printRequests, new SMWQuery(), $results, smwfGetStore() );
```
Comment (JeroenDeDauw) [inline SCQ_CompoundAsk.php]: $queryResult

File: SCQ_CompoundAsk.php [hunk 5]
```diff
@@ -0,0 +1,143 @@
...
+        }
+
+        foreach ( $queryParams as $param ) {
+            $subQueryParams = SCQQueryProcessor::getSubParams( $param );
+
+            if ( array_key_exists( 'format', $other_params ) && !array_key_exists( 'format', $subQueryParams ) ) {
+                $subQueryParams['format'] = $other_params['format'];
+            }
+
+            $next_result = SCQQueryProcessor::getQueryResultFromFunctionParams(
```
Comment (JeroenDeDauw) [inline SCQ_CompoundAsk.php]: #nextResult

File: SCQ_CompoundAsk.php [hunk 6]
```diff
@@ -0,0 +1,143 @@
...
+            SMWQueryProcessor::addThisPrintout( $printRequests, $other_params );
+            $other_params = SMWQueryProcessor::getProcessedParams( $other_params, $printRequests );
+        }
+
+        if ( $this->getMain()->getPrinter() instanceof \ApiFormatXml ) {
+            $outputFormat = 'xml';
+        }
+
+        $this->addQueryResult( $query_result, $outputFormat );
+    }
```
Comment (JeroenDeDauw) [inline SCQ_CompoundAsk.php]: Probably more things you can split out of this method - it's quite long. Not immediately obvious what, as the control flow and side effects are not trivial, all the more reason to split things and simplify the code.

**Comment by PeterTheOne** [comment]
@yaronkoren Format? Yes, it works.

@JeroenDeDauw Most of the code is a straight copy from https://github.com/wikimedia/mediawiki-extensions-SemanticCompoundQueries/blob/master/SCQ_QueryProcessor.php#L30.

**Comment by yaronkoren** [comment]
@PeterTheOne - okay, great. By "formats" I meant "result formats".

I don't like the name "Ask" in the class name. "ask" is of course the name of the parser function, but it's not really a synonym for "SMW query". Maybe the class name could give some indication that it relates to Ajax? Cargo uses the term "deferred query" - maybe this class could use something like that?

**Comment by JeroenDeDauw** [comment]
> Most of the code is a straight copy

That makes me more grumpy about it :)

> I don't like the name [...]

Looking at it, me neither. Definitely needs to have `Api` in it. The SMW class name itself does not, but that is because it is in an Api NS (so actually the class FQN does have Api in it).

How about `CompoundQueryApi`? (Compound Query already implies Ask query, at least in the context of this extension.)

**Comment by yaronkoren** [comment]
Oh! Now I get it - this just adds an API action. (I didn't look at the code that carefully before.) In that case, yes, "CompoundQueryAPI" (or "...Api") makes much more sense as a class name. And similarly, the action itself should probably be named "compoundquery".

In the "examples" section, you have stuff that basically looks like a call to '#compound_query' within the URL. Will those really work? Don't the question marks need to be escaped, and so forth?

**Comment by yaronkoren** [comment]
Oops, sorry, I didn't realize the previous comment was from @JeroenDeDauw , not @PeterTheOne . The 2nd paragraph was addressed to Peter.

**Comment by PeterTheOne** [comment]
> That makes me more grumpy about it :)

:)

About the name: I used "Ask" because it is similar to the Ask Api, but that was just an idea. I like `CompoundQueryApi` for the class name and `compoundquery` for the action name.

> In the "examples" section, you have stuff that basically looks like a call to '#compound_query' within the URL. Will those really work? Don't the question marks need to be escaped, and so forth?

It did work for me in chrome, but I could also just encode it to be safe.

File: SCQ_CompoundAsk.php (hunk 2 above)
Comment (PeterTheOne) [inline SCQ_CompoundAsk.php]: fixed in b335c5a.

File: SCQ_CompoundAsk.php (hunk 3 above)
Comment (PeterTheOne) [inline SCQ_CompoundAsk.php]: done in b335c5a.

File: SCQ_CompoundAsk.php (hunk 4 above)
Comment (PeterTheOne) [inline SCQ_CompoundAsk.php]: done in b335c5a.

File: SCQ_CompoundAsk.php (hunk 5 above)
Comment (PeterTheOne) [inline SCQ_CompoundAsk.php]: done in b335c5a.

File: SCQ_CompoundAsk.php (hunk 6 above)
Comment (PeterTheOne) [inline SCQ_CompoundAsk.php]: done in b335c5a.

**Comment by PeterTheOne** [comment]
> I like `CompoundQueryApi` for the class name and `compoundquery` for the action name.

I changed the name in b335c5a.

I also fixed the encoding thing.

File: SCQ_CompoundAsk.php (hunk 1 above)
Comment (PeterTheOne) [inline SCQ_CompoundAsk.php]: Haven't done anything about this.. Can you explain why this is not nice design?

**Comment by JeroenDeDauw** [comment]
/me whispers to @yaronkoren to move his SMW extensions to GitHub already ;p (I'm happy to help with that)

**Comment by yaronkoren** [comment]
@PeterTheOne - great, thank you! I just merged your changes into the code. I did it manually - I don't know if there's a way to do a merge directly from GitHub, but in any case that gave me more control - I actually split this change up into three patches, plus I slightly changed your API examples, to make them more "semantic". Please let me know if there are any issues with my changes - I didn't actually test it.

**Comment by yaronkoren** [comment]
@JeroenDeDauw - maybe moving them to GitHub might be a good idea, for consistency. We're really just talking about two extensions, I think: this one and Semantic Drilldown. (There's also Semantic Internal Objects, but I don't know if anyone is using it any more.)

This may be a bigger discussion, but: do you think there's any chance that you, James and/or someone else could take over the development of SCQ and SD? I haven't put much effort into either one recently, and perhaps with a move it also makes sense to transfer over control. (Perhaps I should ask @PeterTheOne this same question!)

**Comment by PeterTheOne** [comment]
@yaronkoren Thanks for the merge! :)

**Comment by JeroenDeDauw** [comment]
@yaronkoren You could just have gotten the commits from the branch here on GitHub and then pushed those to Gerrit.

If you already have a clone of gerrit, then that's basically

* git remote add gh https://github.com/wikimedia/mediawiki-extensions-SemanticCompoundQueries.git
* git fetchall
* git checkout `<name of branch on github>`
* `git review` or whatever else you do normally for gerrit

**Comment by JeroenDeDauw** [comment]
@yaronkoren I cannot take over maintenance of any extension. However if it is on GH, it will be easier for me to provide some amount of support, which I expect to be the case for others as well.

**Comment by PeterTheOne** [comment]
> However if it is on GH, it will be easier for me to provide some amount of support, [...]

:+1:

**Comment by JeroenDeDauw** [comment]
@yaronkoren shall I create a copy of the repos on GitHub? You won't have to do anything on that side, though if you want the Gerrit copies to be removed, you'll have to go through whatever relevant WMF processes.

**Comment by yaronkoren** [comment]
There's already a copy on GitHub, no?

**Comment by JeroenDeDauw** [comment]
I meant a copy you can actually write things to without WMF people getting mad about it

**Comment by JeroenDeDauw** [comment]
@yaronkoren I've created such a repo here https://github.com/SemanticMediaWiki/SemanticCompoundQueries

If you confirm you're OK with it, I'll list it as canonical repository, and poke people such as TWN to use it, and to have the Gerrit one removed.

**Comment by yaronkoren** [comment]
@JeroenDeDauw - sorry for not responding before. Yes, that's fine with me. And if this one works out well, it probably makes sense to do the same thing for Semantic Drilldown.

**Comment by JeroenDeDauw** [comment]
Ok, you should now have admin rights on that repo

**Pull request closed without merging**